source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,232 @@
+
+#include <iomanip>
+#include <sstream>
+
+#include "envoy/network/address.h"
+
+#include "common/common/empty_string.h"
+#include "common/network/address_impl.h"
+#include "common/network/utility.h"
+
+// TODO(abaptiste): add fuzzing tests for DNS message parsing
+#include "ares.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+void BaseDnsRecord::serializeName(Buffer::OwnedImpl& output) {
+  // Iterate over a name e.g. "www.domain.com" once and produce a buffer containing each name
+  // segment prefixed by its length
+  static constexpr char SEPARATOR('.');
+
+  size_t last = 0;
+  size_t count = name_.find_first_of(SEPARATOR);
+  auto iter = name_.begin();
+
+  while (count != std::string::npos) {
+    count -= last;
+    output.writeBEInt<uint8_t>(count);
+    for (size_t i = 0; i < count; i++) {
+      output.writeByte(*iter);
+      ++iter;
+    }
+
+    // periods are not serialized. Skip to the next character
+    if (*iter == SEPARATOR) {
+      ++iter;
+    }
+
+    // Move our last marker to the first position after where we stopped. Search for the next name
+    // separator
+    last += count;
+    ++last;
+    count = name_.find_first_of(SEPARATOR, last);
+  }
+
+  // Write the remaining segment prepended by its length
+  count = name_.size() - last;
+  output.writeBEInt<uint8_t>(count);
+  for (size_t i = 0; i < count; i++) {
+    output.writeByte(*iter);
+    ++iter;
+  }
+
+  // Terminate the name record with a null byte
+  output.writeByte(0x00);
+}
+
+// Serialize a DNS Query Record
+void DnsQueryRecord::serialize(Buffer::OwnedImpl& output) {
+  serializeName(output);
+  output.writeBEInt<uint16_t>(type_);
+  output.writeBEInt<uint16_t>(class_);
+}
+
+DnsQueryContextPtr DnsMessageParser::createQueryContext(Network::UdpRecvData& client_request) {
+  DnsQueryContextPtr query_context = std::make_unique<DnsQueryContext>(
+      client_request.addresses_.local_, client_request.addresses_.peer_);
+
+  query_context->parse_status_ = parseDnsObject(query_context, client_request.buffer_);
+  if (!query_context->parse_status_) {
+    ENVOY_LOG(error, "Unable to parse query buffer from '{}' into a DNS object.",
+              client_request.addresses_.peer_->ip()->addressAsString());
+  }
+
+  return query_context;
+}
+
+bool DnsMessageParser::parseDnsObject(DnsQueryContextPtr& context,
+                                      const Buffer::InstancePtr& buffer) {
+  static constexpr uint64_t field_size = sizeof(uint16_t);
+  size_t available_bytes = buffer->length();
+  uint64_t offset = 0;
+  uint16_t data;
+  DnsQueryParseState state{DnsQueryParseState::Init};
+
+  header_ = {};
+  while (state != DnsQueryParseState::Finish) {
+    // Ensure that we have enough data remaining in the buffer to parse the query
+    if (available_bytes < field_size) {
+      ENVOY_LOG(error,
+                "Exhausted available bytes in the buffer. Insufficient data to parse query field.");
```

Comment:
This can log spam, make this a debug log. This should probably have a stat also which can be done in a follow up. Same below.

Reply:
ACK.  Will add a stat for this for a future PR.